playground/src/csp-hash-calculator/csp-hash-calculator.js

```diff
@@ -0,0 +1,174 @@
+// Copyright 2018 The AMPHTML Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/* eslint-disable new-cap */
+import {calculateHash} from '@ampproject/toolbox-script-csp';
+
+export function createCspHashCalculator(editor) {
+  return new CspHashCalculator(editor);
+}
+
+class CspHashCalculator {
+  constructor(editor) {
+    this.editor = editor;
+    this.head = {from: null, to: null};
+    this.metaAmpScriptSrc = {from: null, to: null};
+
+    // Populated while parsing the documents. Contains the DOM strings either
+    // in the form of `<script target="amp-script">...</script>`
+    this.inlineScripts = [];
+
+    // Stores all hashes that are currently added to the document
+    // to prevent update loops
+    this.createdHashes = [];
+  }
+
+  update() {
+    // Throw away previously parsed scripts and get them from the document
+    this.inlineScripts = [];
+    this._parseDocument();
+
+    if (!this.inlineScripts.length) {
+      return;
+    }
+
+    // If there is no existing meta[name="amp-script-src"] create one
+    // and return as this triggers a new update anyway
+    if (!this.metaAmpScriptSrc.to) {
+      this.editor.replaceRange(`${this._getMetaAmpScriptSrc()}\n`, {
+        line: this.head.to.line,
+        ch: 0,
+      });
+      this.editor.codeMirror.indentLine(this.head.to.line);
+      return;
+    }
+
+    let existingHashes = this._getExistingHashes();
+    let inlineHashes = this._getInlineHashes();
+
+    // Check if any inline hash has updated, if no hashes are left
+    // meta[name="amp-script-src"] is up to date
+    inlineHashes = inlineHashes.filter(
+      (hash) => !existingHashes.includes(hash)
+    );
+    if (!inlineHashes.length) {
+      return;
+    }
+
+    // Now filter out hashes that have previously been added by
+    // the calculator and store the newly created ones
+    existingHashes = existingHashes.filter(
+      (hash) => !this.createdHashes.includes(hash)
+    );
+    this.createdHashes = inlineHashes;
+
+    this.editor.replaceRange(
+      `${this._getMetaAmpScriptSrc([...inlineHashes, ...existingHashes])}`,
+      this.metaAmpScriptSrc.from,
+      this.metaAmpScriptSrc.to
+    );
+  }
+
+  _getExistingHashes() {
+    const html = this.editor.codeMirror.getRange(
+      this.metaAmpScriptSrc.from,
+      this.metaAmpScriptSrc.to
+    );
+    return html.match(/(?<=content=").*?(?=")/gms)[0].split(' ');
```

Comment:
This fails `npm run build:playground`:

```
ERROR in main.2eaca583ec3a65d6b302.js:51525 from closure-compiler: Cannot convert ECMASCRIPT_2018 feature "RegExp Lookbehind" to targeted output language.
      return html.match(/(?<=content=")[\s\S]*?(?=")/gm)[0].split(' ');
```

Can you also add this to the package.json (this got lost during the migration):

```
 "browserslist": [
   "last 2 versions", "not dead"
  ],
```